### PR TITLE
[spirv] Support struct as kernel argument

### PR DESCRIPTION
**Pull request by lin-hitonami**
Only `is_array` is used in `KernelContextAttribute.args()` so we don't need to set the rest of them. We will remove `ArgAttribute` as soon as `is_array` is no longer used.

Will refactor `compile_args_struct` to let it translate the CHI-IR struct type to spirv struct type without using tinyir after we get rid of the extra_args.

### Summary
🔄🆕🛠️

This pull request improves the argument handling and type translation logic for the SPIR-V backend. It simplifies the `KernelContextAttributes` constructor, refactors the `TaskCodegen` class, adds a new method `from_taichi_type` to the `IRBuilder` class, and fixes a test failure in `test_argument.py`. It also introduces some utility functions for aligning and translating struct types.

> _Sing, O Muse, of the valiant codegen heroes_
> _Who tamed the fiery SPIR-V with skill and craft_
> _They forged the `IRBuilder` with wise methods_
> _To shape the types from Taichi as they wished_

### Walkthrough
*  Simplify and fix the argument handling logic for the SPIR-V backend ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L63-R63), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L102-R94), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L120-R115), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L142), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L569-R570), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L576-R578), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000R585-R588), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L2193-R2202), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-259a89b8645aab6065eac94864cd4930be7ae91da5cb2e15307ae5fa41015962R473-R498), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-b87ff864662b493b7a77265f0abfa44586aab4be0908e951270117f6ba9b7221R229-R232), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-f2d9c4a8c504d2073549cde99b31a733a2f39bba4ce9f062cbaa09e1be17aa87L237-R239))
  * Remove the unused fields of `ArgAttributes` in `KernelContextAttributes` constructor in `kernel_utils.cpp` to reduce memory footprint and complexity ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L63-R63))
  * Fix the bug of calculating the size of array arguments when physical storage buffer is not supported in `KernelContextAttributes` constructor in `kernel_utils.cpp` to match the buffer size used by the codegen ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L102-R94))
  * Remove the redundant code of checking the argument offsets and size with the kernel type in `KernelContextAttributes` constructor in `kernel_utils.cpp` to avoid overhead and errors when the kernel type is not available or compatible ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L120-R115), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-0933633c5350c8a0e64f59046da02dcafcabe69694bb200d938b8bd415bcbca0L142))
  * Use the argument type from `KernelContextAttributes` instead of the argument attributes in `TaskCodegen::visit` and `TaskCodegen::compile_args_struct` in `spirv_codegen.cpp` to make the code more consistent and robust ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L569-R570), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L2193-R2202))
  * Use the `IRBuilder::from_taichi_type` method to get the SPIR-V type for a given Taichi type in `TaskCodegen::visit` in `spirv_codegen.cpp` to abstract away the details of the SPIR-V type system and handle different cases of pointer and struct types ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L576-R578))
  * Skip the load operation for `ArgLoadStmt` if the `create_load` flag is false in `TaskCodegen::visit` in `spirv_codegen.cpp` to support the case where the argument is passed as a pointer without dereferencing it ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000R585-R588))
  * Add the `translate_ti_type` function to `spirv_types.cpp` and `spirv_types.h` to reuse the existing logic of translating primitive types and extend it to handle pointer and struct types ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-259a89b8645aab6065eac94864cd4930be7ae91da5cb2e15307ae5fa41015962R473-R498), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-b87ff864662b493b7a77265f0abfa44586aab4be0908e951270117f6ba9b7221R229-R232))
  * Modify the `test_struct_arg` function in `test_argument.py` to use a different struct type that is compatible with the SPIR-V backend to fix a test failure caused by the previous changes ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-f2d9c4a8c504d2073549cde99b31a733a2f39bba4ce9f062cbaa09e1be17aa87L237-R239))
* Add a new method `IRBuilder::from_taichi_type` to the `IRBuilder` class that returns the SPIR-V type for a given Taichi type ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-5274ca34a1156d70b415ce034aeea8430182b5b15b95b3cef9efa79661028e15R331-R351), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-e9e785201af78fa53c937a623f0bd851161ac2064d92383c05384f786cea9cd9L340-R343))
  * Define the `from_taichi_type` method in `spirv_ir_builder.cpp` to abstract away the details of the SPIR-V type system and handle different cases of pointer and struct types ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-5274ca34a1156d70b415ce034aeea8430182b5b15b95b3cef9efa79661028e15R331-R351))
  * Declare the `from_taichi_type` method and rename the existing `get_primitive_type` method in `spirv_ir_builder.h` to update the interface of the `IRBuilder` class ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-e9e785201af78fa53c937a623f0bd851161ac2064d92383c05384f786cea9cd9L340-R343))
* Add a new function `align_up` to the `type_utils.h` file that returns the smallest multiple of a given alignment that is greater than or equal to a given value ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1df7532afc2540435b46bd5831a4b6bd11bb1ead5e1fd3607e02fccf9dab112cR21-R24))
  * Define the `align_up` function in `type_utils.h` to provide a utility function for aligning memory offsets and sizes ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-1df7532afc2540435b46bd5831a4b6bd11bb1ead5e1fd3607e02fccf9dab112cR21-R24))
* Refactor the `get_struct_type_with_data_layout` method of `GfxRuntime` to use a new helper function `get_struct_type_with_data_layout_impl` that handles the recursive logic of translating and aligning struct types ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-5de6a93ab8620f5ececa55c3e762efa567d428c85058593ab49bb641f0c4c2fcL792-R860), [link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-a023ad94c9af40b9b4b391fc6ec0d96e9c7f88744d196bf782c51086da96d99aR136-R140))
  * Define the `get_struct_type_with_data_layout_impl` function in `runtime.cpp` to handle the recursive logic of translating and aligning struct types ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-5de6a93ab8620f5ececa55c3e762efa567d428c85058593ab49bb641f0c4c2fcL792-R860))
  * Declare the `get_struct_type_with_data_layout_impl` function in `runtime.h` to make it visible to the `runtime.cpp` file ([link](https://github.com/taichi-dev/taichi/pull/7788/files?diff=unified&w=0#diff-a023ad94c9af40b9b4b391fc6ec0d96e9c7f88744d196bf782c51086da96d99aR136-R140))

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7849
* #7809
* __->__ #7788

### CODE REVIEW

File: taichi/codegen/spirv/kernel_utils.cpp [hunk 1]
```diff
@@ -99,8 +90,8 @@ KernelContextAttributes::KernelContextAttributes(
     for (int i = 0; i < vec->size(); ++i) {
       auto &attribs = (*vec)[i];
       const size_t dt_bytes =
-          (attribs.is_array && !is_ret && has_buffer_ptr)
-              ? sizeof(uint64_t)
+          (attribs.is_array && !is_ret)
+              ? (has_buffer_ptr ? sizeof(uint64_t) : sizeof(uint32_t))
```
Comment (bobcao3) [inline taichi/codegen/spirv/kernel_utils.cpp:94]: What's `sizeof(uint32_t)` for here?

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 2]
```diff
@@ -566,21 +566,26 @@ class TaskCodegen : public IRVisitor {
...
+      bool has_buffer_ptr =
+          caps_->get(DeviceCapability::spirv_has_physical_storage_buffer);
+      const auto val_type = ir_->from_taichi_type(arg_type, has_buffer_ptr);
       spirv::Value buffer_val = ir_->make_value(
           spv::OpAccessChain,
           ir_->get_pointer_type(val_type, spv::StorageClassUniform),
           get_buffer_value(BufferType::Args, PrimitiveType::i32),
           ir_->int_immediate_number(ir_->i32_type(), arg_id));
       buffer_val.flag = ValueKind::kVariablePtr;
+      if (!stmt->create_load) {
```
Comment (bobcao3) [inline taichi/codegen/spirv/spirv_codegen.cpp:585]: What's this? Am I understanding this correctly: `ArgLoadStmt` now has two types, one actually loads argument, the other generates a pointer (basically a special GEP?)

Comment (lin-hitonami) [reply taichi/codegen/spirv/spirv_codegen.cpp:585]: Yes. I plan to split `ArgLoad` into an `ArgAddr` wihch just generates a GEP and a `GlobalLoad` in the future.

File: taichi/codegen/spirv/kernel_utils.cpp (hunk 1 above)
Comment (lin-hitonami) [reply taichi/codegen/spirv/kernel_utils.cpp:94]: When `has_buffer_ptr` is false, we need a placeholder argument. It was the element type of the pointer before. In this PR I changed it to u32 for simplicity.

Comment (lin-hitonami) [reply taichi/codegen/spirv/kernel_utils.cpp:94]: BTW we no longer use this function to calculate the offset of the arguments anymore. This function still calculates the offset of rets. We will remove them along with `ArgAttribute` and `RetAttribute` as soon as the refactor completes.

File: taichi/codegen/spirv/spirv_ir_builder.cpp [hunk 3]
```diff
@@ -328,6 +328,27 @@ SType IRBuilder::get_primitive_type(const DataType &dt) const {
...
 }
 
+SType IRBuilder::from_taichi_type(const DataType &dt, bool has_buffer_ptr) {
+  if (dt->is<PrimitiveType>()) {
+    return get_primitive_type(dt);
+  } else if (dt->is<PointerType>()) {
+    if (has_buffer_ptr) {
+      return t_uint64_;
+    } else {
+      return t_uint32_;
```
Comment (bobcao3) [inline taichi/codegen/spirv/spirv_ir_builder.cpp:338]: ditto, what's this?

Comment (lin-hitonami) [reply taichi/codegen/spirv/spirv_ir_builder.cpp:338]: Same as above. When has_buffer_ptr is false, we need a placeholder argument. It was the element type of the pointer before. In this PR I changed it to u32 for simplicity.

**Review by bobcao3 — APPROVED** [review]
LGTM

**Pull request closed without merging**